File: Python/40. Combination Sum II.py
class Solution:
    def combinationSum2(self, candidates: list[int], target: int) -> list[list[int]]:
        candidates.sort()
        result = []
        sub_arr = []

        def backtracking(arr: list[int], target: int, sub_arr: list[int]):
            i = 0
            while i < len(arr):
                if arr[i] > target:
                    break
                if arr[i] == target:
                    result.append(sub_arr + [arr[i]])
                    break

                backtracking(arr[i+1:], target - arr[i], sub_arr + [arr[i]])

                # For prevent duplicate
                j = 1
                while i+j < len(arr) and arr[i+j] == arr[i]:
                    j += 1
                i += j
                
        backtracking(candidates, target, sub_arr)
        
        return result

File: Javascript/40. Combination Sum II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    let result = []
    candidates.sort((a,b) => a - b)
    
    let temp = []

    function backtracking(left, res) {
        if (res < 0) {
            return ;
        }
        if (res == 0) {
            result.push([...temp])
            return 
        }

        let i = left
        while (i < candidates.length) {
            temp.push(candidates[i])
            backtracking(i+1, res - candidates[i])
            temp.pop()
            let j = 1
            while (i+j < candidates.length && candidates[i] == candidates[i+j]) j++;
            i += j
        }
        return
    }

    backtracking(0, target)
    return result
};